File: android/src/main/kotlin/com/klarna/inapp/sdk/flutter_klarna_inapp_sdk/postpurchase/KlarnaPostPurchaseSDKHandler.kt
package com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.postpurchase

import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.core.handler.BaseMethodHandler
import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.postpurchase.manager.KlarnaPostPurchaseSDKManager
import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.postpurchase.method.KlarnaPostPurchaseSDKMethod
import io.flutter.plugin.common.MethodChannel

internal object KlarnaPostPurchaseSDKHandler :
    BaseMethodHandler<KlarnaPostPurchaseSDKMethod>(KlarnaPostPurchaseSDKMethod.Parser) {

    val postPurchaseSDKManagerMap = mutableMapOf<Int, KlarnaPostPurchaseSDKManager>()

    override fun onMethod(method: KlarnaPostPurchaseSDKMethod, result: MethodChannel.Result) {
        when (method) {
            is KlarnaPostPurchaseSDKMethod.Create -> create(method, result)
            is KlarnaPostPurchaseSDKMethod.Initialize -> initialize(method, result)
            is KlarnaPostPurchaseSDKMethod.Destroy -> destroy(method, result)
            is KlarnaPostPurchaseSDKMethod.RenderOperation -> renderOperation(method, result)
            is KlarnaPostPurchaseSDKMethod.AuthorizationRequest -> authorizationRequest(
                method,
                result
            )
        }
    }

    private fun create(
        method: KlarnaPostPurchaseSDKMethod.Create,
        result: MethodChannel.Result
    ) {
        val manager = getManager(method)
        if (manager == null) {
            val newManager = KlarnaPostPurchaseSDKManager(method.id)
            newManager.create(method, result)
            postPurchaseSDKManagerMap[method.id] = newManager
        } else {
            getManager(method)?.create(method, result)
        }
    }

    private fun initialize(
        method: KlarnaPostPurchaseSDKMethod.Initialize,
        result: MethodChannel.Result
    ) {
        getManager(method)?.initialize(method, result)
            ?: KlarnaPostPurchaseSDKManager.notInitialized(result)
    }

    private fun authorizationRequest(
        method: KlarnaPostPurchaseSDKMethod.AuthorizationRequest,
        result: MethodChannel.Result
    ) {
        getManager(method)?.authorizationRequest(method, result)
            ?: KlarnaPostPurchaseSDKManager.notInitialized(result)
    }

    private fun renderOperation(
        method: KlarnaPostPurchaseSDKMethod.RenderOperation,
        result: MethodChannel.Result
    ) {
        getManager(method)?.renderOperation(method, result)
            ?: KlarnaPostPurchaseSDKManager.notInitialized(result)
    }

    private fun destroy(method: KlarnaPostPurchaseSDKMethod.Destroy, result: MethodChannel.Result) {
        getManager(method)?.destroy(method, result)
            ?: KlarnaPostPurchaseSDKManager.notInitialized(result)
        postPurchaseSDKManagerMap.remove(method.id)
    }

    private fun getManager(method: KlarnaPostPurchaseSDKMethod): KlarnaPostPurchaseSDKManager? {
        return postPurchaseSDKManagerMap[method.id]
    }
}

File: ios/Classes/PostPurchase/KlarnaPostPurchaseSDKHandler.swift
import Flutter
import WebKit
import KlarnaMobileSDK

class KlarnaPostPurchaseSDKHandler: BaseMethodHandler<KlarnaPostPurchaseSDKMethod> {
    
    private var postPurchaseSDKManagerMap = [Int:KlarnaPostPurchaseSDKManager]()
    
    init() {
        super.init(parser: KlarnaPostPurchaseSDKMethods.Parser())
    }
    
    override func onMethod(method: KlarnaPostPurchaseSDKMethod, result: @escaping FlutterResult) {
        switch method {
        case is KlarnaPostPurchaseSDKMethods.Create:
            create(method: method as! KlarnaPostPurchaseSDKMethods.Create, result: result)
        case is KlarnaPostPurchaseSDKMethods.Initialize:
            initialize(method: method as! KlarnaPostPurchaseSDKMethods.Initialize, result: result)
        case is KlarnaPostPurchaseSDKMethods.Destroy:
            destroy(method: method as! KlarnaPostPurchaseSDKMethods.Destroy, result: result)
        case is KlarnaPostPurchaseSDKMethods.RenderOperation:
            renderOperation(method: method as! KlarnaPostPurchaseSDKMethods.RenderOperation, result: result)
        case is KlarnaPostPurchaseSDKMethods.AuthorizationRequest:
            authorizationRequest(method: method as! KlarnaPostPurchaseSDKMethods.AuthorizationRequest, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func create(method: KlarnaPostPurchaseSDKMethods.Create, result: @escaping FlutterResult) {
        if let manager = getManager(method: method) {
            manager.create(method: method, result: result)
        } else {
            let manager = KlarnaPostPurchaseSDKManager(id: method.id)
            manager.create(method: method, result: result)
            postPurchaseSDKManagerMap[method.id] = manager
        }
    }
    
    private func initialize(method: KlarnaPostPurchaseSDKMethods.Initialize, result: @escaping FlutterResult) {
        getManager(method: method)?.initialize(method: method, result: result) ?? KlarnaPostPurchaseSDKManager.notInitialized(result: result)
    }
    
    private func authorizationRequest(method: KlarnaPostPurchaseSDKMethods.AuthorizationRequest, result: @escaping FlutterResult) {
        getManager(method: method)?.authorizationRequest(method: method, result: result) ?? KlarnaPostPurchaseSDKManager.notInitialized(result: result)
    }
    
    private func renderOperation(method: KlarnaPostPurchaseSDKMethods.RenderOperation, result: @escaping FlutterResult) {
        getManager(method: method)?.renderOperation(method: method, result: result) ?? KlarnaPostPurchaseSDKManager.notInitialized(result: result)
    }
    
    private func destroy(method: KlarnaPostPurchaseSDKMethods.Destroy, result: @escaping FlutterResult) {
        getManager(method: method)?.destroy(method: method, result: result)
        postPurchaseSDKManagerMap.removeValue(forKey: method.id)
    }
    
    private func getManager(method: KlarnaPostPurchaseSDKMethod) -> KlarnaPostPurchaseSDKManager? {
        return postPurchaseSDKManagerMap[method.id]
    }
}
